app/lib/package/export_api_to_cdn.dart

```diff
@@ -0,0 +1,239 @@
+  final gzippedBytes = gzip.encode(rawBytes);
+
+  for (final objectName in _cdnObjectNames(package)) {
+    final info = await bucket.tryInfo(objectName);
+    // Skip upload if was updated recently, and the bytes length and md5 hash matches.
+    if (info != null &&
+        clock.now().difference(info.updated).inHours <= 12 &&
+        info.length == gzippedBytes.length) {
+      final md5Hash = md5.convert(gzippedBytes).bytes;
+      if (info.md5Hash.length == md5Hash.length &&
+          info.md5Hash.whereIndexed((i, e) => md5Hash[i] == e).length ==
+              info.md5Hash.length) {
+        continue;
+      }
+    }
+
+    await uploadWithRetry(
+      bucket,
+      objectName,
+      gzippedBytes.length,
+      () => Stream.value(gzippedBytes),
+      metadata: ObjectMetadata(
+        contentType: 'application/json',
+        contentEncoding: 'gzip',
+      ),
+    );
+  }
+}
+
+Future<void> _deletesPackageFromBucket(String package, Bucket bucket) async {
+  for (final objectName in _cdnObjectNames(package)) {
+    await bucket.tryDelete(objectName);
+  }
+}
+
+Future<Map<String, bool>> _queryRecentlyUpdated(
+    DateTime lastQueryStarted) async {
+  final updatedThreshold = lastQueryStarted.subtract(Duration(minutes: 5));
+  final results = <String, bool>{};
+  void addResult(String pkg, bool isVisible) {
+    final current = results[pkg];
+    // allow override only into the non-visible direction
+    if (current ?? true) {
+      results[pkg] = isVisible;
+    }
+  }
+
+  final q1 = dbService.query<ModeratedPackage>()
+    ..filter('moderated >=', updatedThreshold)
+    ..order('-moderated');
+  await for (final s in q1.run()) {
+    addResult(s.name!, false);
+  }
+
+  final q2 = dbService.query<Package>()
+    ..filter('updated >=', updatedThreshold)
+    ..order('-updated');
+  await for (final p in q2.run()) {
+    addResult(p.name!, p.isVisible);
+  }
+
+  return results;
+}
+
+/// Deletes obsolete files on the CDN [bucket].
+Future<void> gcPackageApiCdn(Bucket bucket) async {
+  final versions = <String>{};
+
+  // Objects in the bucket are stored under the following pattern:
+  //   `current/api/<package>`
+  //   `<runtimeVersion>/api/<package>`
+  // Thus, we list with `/` as delimiter and get a list of runtimeVersions
+  await for (final d in bucket.list(prefix: '', delimiter: '/')) {
+    if (!d.isDirectory) {
+      _logger.warning('bucket should not contain any top-level object');
+      continue;
+    }
+
+    // Remove trailing slash from object prefix, to get a runtimeVersion
+    assert(d.name.endsWith('/'));
+    final rtVersion = d.name.substring(0, d.name.length - 1);
+    if (runtimeVersionPattern.matchAsPrefix(rtVersion) == null) {
+      continue;
+    }
+
+    // Check if the runtimeVersion should be GC'ed
+    if (shouldGCVersion(rtVersion)) {
+      versions.add(rtVersion);
+    }
+  }
+
+  for (final v in versions) {
+    await deleteBucketFolderRecursively(bucket, '$v/', concurrency: 4);
```

Comment:
Just an aside (not for this CL) perhaps we can use batch requests https://cloud.google.com/storage/docs/batch to delete files faster.